fix: fix client id scheme (#1)
awoie/OpenID4VP: openid-4-verifiable-presentations-1_0.md

## Verifier Metadata Management {#client_metadata_management}

> If the Client is not pre-registered with the Wallet, the `client_id` MUST be a URI.

Reviewer: We should be prepared for the eventuality that people may say that this is an unnecessary change, since for the x509_san_dns case, the client_id currently is a DNS name and not a URI.

What URI value are we suggesting that people use after this PR is merged when they were previously using a DNS name with x509_san_dns?

Author: That makes sense.

> The Wallet SHOULD verify the provided `redirect_uri` or `response_uri` belongs to the Client.

Reviewer: ```suggestion
The Wallet MUST verify the provided `redirect_uri` or `response_uri` belongs to the Client.
(Note that it is currently unspecified how to achieve this for DIDs.)
```

If the Authorization Request Object is signed, the Wallet MUST authenticate the Request Object against the `client_id` contained in the Request Object.

> If the Authorization Request is signed, the `jwks` parameter in the `client_metadata` parameter MUST NOT be used to verify the signature of the Request Object.

Reviewer: ```suggestion
If the Authorization Request is signed, any `jwks` parameter in the `client_metadata` parameter MUST NOT be used to verify the signature of the Request Object.
```

> TODO: include some text that `client_metadata` parameter does not take precedence over client metadata provided by other mechanisms that are more trusted (except for ephemeral encryption keys), e.g., DCR, OpenID Federation

Author: @selfissued: do you have a better way to frame this?

Reviewer: Why are we bringing this topic up?  I think it's more likely to generate a comment stream and objections than to get the proposal accepted.  I'd delete it for now.  But remember to add something like it after the proposal is adopted.

Reviewer: I'd still suggest deleting the precedence statement.

> TODO: include some text that `client_metadata` parameter does not take precedence over client metadata provided by other mechanisms that are more trusted (except for ephemeral encryption keys), e.g., DCR, OpenID Federation

Reviewer: ```suggestion
```

> If the Client is not pre-registered and if the Authorization Request is unsigned, the Wallet MUST check if the `client_id` matches the `redirect_uri` or `response_uri` parameter.

Author: This is probably an issue for OpenID Federation.

Author: Perhaps we should say instead:
- If the resolved Verifier Metadata parameter does not contain `redirect_uris`, the Wallet MUST check if `client_id` matches the `redirect_uri` or `response_uri` parameter.
- The `client_metadata` parameter MUST not contain the `redirect_uris` parameter.

Reviewer: We need to fix this to allow the OpenID Federation behavior before publishing.  In particular, for the redirect_uris metadata parameter to be in the metadata and for the redirect_uri to be matched against that set.

> If the Client is not pre-registered and if the Authorization Request is unsigned, the Wallet MUST check if the `client_id` matches the `redirect_uri` or `response_uri` parameter.

Reviewer: ```suggestion
If the Client is not pre-registered and if the Authorization Request is unsigned, the Wallet MUST check if the `client_id` matches the `redirect_uri` or `response_uri` parameter and one of the `redirect_uris` values in the resolved metadata.
```

Author: For pre-registered clients or clients where metadata can be resolved through existing mechanisms, the `redirect_uris` can be checked. Most client ID schemes don't support that which is why I added this text for security reasons. It matches what was already in the spec in the respective client ID scheme sections.

Reviewer: Please merge the suggestion above to fix the OpenID Federation issue before publishing.

It is a Wallet-policy decision on how to authenticate the Request Object, whether to accept unsigned Authorization Requests and whether the linkage between the `redirect_uri` or `response_uri` and the Client is verified.

Note: To prevent downgrade attacks, a Client can choose to include certain parameters in the request object or not.

> The following is a list of methods defined by this specification that MAY be used by the Wallet to authenticate the Request Object against the `client_id` in the Request Object and verify the `redirect_uri` or `response_uri` belong to the Client:

Reviewer: ```suggestion
The following is a list of methods defined by this specification that are used by the Wallet to authenticate the Request Object against the `client_id` in the Request Object and verify that the `redirect_uri` or `response_uri` belong to the Client:
```

- Pre-registered: If the Client is pre-registered, the `jwks` Client Metadata parameter provided upon registration MUST be used to verify the signature of the Request Object.

> - X.509 certificates: If the Request Object contains an `x5c`, `x5t`, or `x5t#S256` header parameter, after validating the X.509 certificate chain, the Wallet MUST use the public key from the leaf-certificate of the X.509 certificate chain to verify the signature of the Request Object. In this case, the leaf-certificate MUST include a`dNSName` Subject Alternative Name (SAN) [RFC5280] that matches the FQDN of the `client_id`, or `uniformResourceIdentifier` SAN [RFC5280] that matches the `client_id`. If the Wallet can establish trust in the `client_id` authenticated through the certificate, e.g. because the `client_id` is contained in a list of trusted Client Identifiers, it MAY allow the Client to freely choose the `redirect_uri` or `response_uri` value. If not, the `redirect_uri` or `response_uri` value MUST match the `client_id`.

Reviewer: It's not obvious to me why we're talking about allowing the Client to freely choose the `redirect_uri` or `response_uri` value.  Is this already the case in the spec?  Why would we want this?

Author: Yes, that is the current spec text

Reviewer: I wonder why it talks about that.  I don't understand the motivation.

Author: security related but it is not consistent with the other schemes. that is why i tried to crystalize the commonality and security functions of all client id schemes.

> - DID Resolution: If the `client_id` in the Request Object is a DID, a public key to verify the signature MUST be obtained from the `verificationMethod` property of a DID Document. Since a DID Document can include multiple public keys, a particular public key used to sign the request in question MUST be identified by the `kid` header parameter. To obtain the DID Document, the Wallet MUST use DID Resolution defined by the DID method used by the Verifier.

Reviewer: ```suggestion
- DID Resolution: If the `client_id` in the Request Object is a DID, a public key to verify the signature MUST be obtained from a `verificationMethod` property of a DID Document. Since a DID Document can include multiple public keys, the particular public key used to sign the request MUST be identified by the `kid` header parameter. To obtain the DID Document, the Wallet MUST use DID Resolution as defined by the DID method used by the Verifier.
```

> - Verifier Attestation: If the Request Object contains a Verifier Attestation JWT as defined in Section 10, the `client_id` MUST equal the `sub` claim value in the Verifier Attestation JWT. The Wallet MUST verify the signature of the Request Object using the public key in the `cnf` claim in the Verifier attestation JWT. This serves as proof of possesion of this key. The Wallet MUST validate the signature on the Verifier Attestation JWT. The `iss` claim value of the Verifier Attestation JWT MUST identify a party the Wallet trusts for issuing Verifier Attestation JWTs. If the issuer of the Verifier Attestation JWT adds a `redirect_uris` claim to the Verifier Attestation, the Wallet MUST ensure the `redirect_uri` request parameter value exactly matches one of the `redirect_uris` claim entries.

Reviewer: ```suggestion
- Verifier Attestation: If the Request Object contains a Verifier Attestation JWT as defined in (#verifier_attestation_jwt), the `client_id` MUST equal the `sub` claim value in the Verifier Attestation JWT. The Wallet MUST verify the signature of the Request Object using the public key in the `cnf` claim in the Verifier Attestation JWT. This serves as proof of possesion of this key. The Wallet MUST validate the signature on the Verifier Attestation JWT. The `iss` claim value of the Verifier Attestation JWT MUST identify a party the Wallet trusts for issuing Verifier Attestation JWTs. If the issuer of the Verifier Attestation JWT includes a `redirect_uris` claim in the Verifier Attestation, the Wallet MUST ensure the `redirect_uri` request parameter value exactly matches one of the `redirect_uris` array values.
```

> - OpenID Federation: If the `client_id` in the Request Object is an Entity Identifier defined in OpenID Federation [OpenID.Federation], the processing rules given in [OpenID.Federation] MUST be followed. Automatic Registration as defined in [OpenID.Federation] MUST be used. The Authorization Request MAY also contain a `trust_chain` parameter. The `jwks` parameter of the final Verifier metadata obtained from the Trust Chain after applying the policies, according to [OpenID.Federation], MUST be used to verify the signature of the Request Object.

Reviewer: ```suggestion
- OpenID Federation: If the `client_id` in the Request Object is an Entity Identifier as defined in OpenID Federation [@!OpenID.Federation], the processing rules specified in OpenID Federation MUST be followed. When this is the case, the Entity Configuration for the client is obtained at the path <client_id>`/.well-known/openid-federation`. Automatic Registration as defined in OpenID Federation MUST be used. The Authorization Request MAY also contain a `trust_chain` parameter, which can optimize the process of obtaining metadata. The `jwks` parameter of the resolved Verifier metadata obtained from the Trust Chain after applying the policies, according to OpenID Federation, MUST be used to verify the signature of the Request Object.
```

> Separate specifications or ecosystem regulations MAY define rules complementing the rules defined above, but such rules are out of scope of this specification. 

Reviewer: ```suggestion
Other specifications or ecosystem regulations MAY define rules complementing the rules defined above, but such extensions are out of scope of this specification. 
```

## Verifier Metadata Management {#client_metadata_management}

If the Client is not pre-registered with the Wallet, the `client_id` MUST be a URI.

> The Wallet SHOULD verify the provided `redirect_uri` or `response_uri` belongs to the Client.

Author: @selfissued: it is a SHOULD because for DIDs there is no obvious way how to do that, even with the existing OID4VP text.

Reviewer: I'd make it a MUST.  And then maybe add this comment:
(Note that it is currently unspecified how to achieve this for DIDs.)

If the Authorization Request Object is signed, the Wallet MUST authenticate the Request Object against the `client_id` contained in the Request Object.

> If the Authorization Request is signed, the `jwks` parameter in the `client_metadata` parameter MUST NOT be used to verify the signature of the Request Object.

Author: @selfissued: the above sentence eleviates the need to repeat that a JWK in the request cannot be used to verify the signature of the request which is quite obvious.

TODO: include some text that `client_metadata` parameter does not take precedence over client metadata provided by other mechanisms that are more trusted (except for ephemeral encryption keys), e.g., DCR, OpenID Federation

The Wallet MUST Verifier metadata provided by the `client_metadata` parameter MUST NOT override Verifier metadata by other existing mechanisms, e.g., 

If the Client is not pre-registered and if the Authorization Request is unsigned, the Wallet MUST check if the `client_id` matches the `redirect_uri` or `response_uri` parameter.

It is a Wallet-policy decision on how to authenticate the Request Object, whether to accept unsigned Authorization Requests and whether the linkage between the `redirect_uri` or `response_uri` and the Client is verified.

Note: To prevent downgrade attacks, a Client can choose to include certain parameters in the request object or not.

> The following is a list of methods defined by this specification that MAY be used by the Wallet to authenticate the Request Object against the `client_id` in the Request Object and verify the `redirect_uri` or `response_uri` belong to the Client:

Author: @selfissued: does the above MAY express that even if the methods below contain a MUST, they only apply in case the Wallet chose to apply one of these rules?

Reviewer: I suggested changing "MAY be used" to "are used".

Author: That makes sense

> The following is a list of methods defined by this specification that MAY be used by the Wallet to authenticate the Request Object against the `client_id` in the Request Object and verify the `redirect_uri` or `response_uri` belong to the Client:

Author: @selfissued: We could move these methods to a dedicated security considerations section on "Authenticating requests for non-pre-registered Clients" and "Verify linkage between Client ID and Redirect/ Response URI", or we just keep it in the Verifier Metadata Management chapter. What do you think would be better?

Reviewer: Normative requirements belong in the main body of the specification - not the Security Considerations section.

Author: Ok, let's do that in the first round and file another issue to move the parts to other sections if needed.
